### DISCUSSION TITLE
Video playback on Linux not present

### DISCUSSION THREAD

**Discussion by Smorty100** [ppy/osu#14364] [Q&A]
I am probably doing something wrong, but on every Linux install which I have (all ubuntu based), none could play any video back. I checked that these beatmaps have video via osu!stable on Linux (via Lutris, wine) and it seems that it just won't play anything. Yes I have a systemwide ffmpeg install on every linux install. I don't get any kind of error message from lazer, it just shows the default background (by the song) as if there wasn't any video to begin with. Yes I specifically clicked on "download with song" in osu!direct.

**Comment by gagahpangeran** [comment]
Do you have ffmpeg installed on your system. Currently we don't build the appimage with ffmpeg included for linux.

Also please provide logs.

**Reply by bdach** [reply]
The user has already claimed that they have a system-wide ffmpeg.

@smorty100 please provide output of `ffmpeg --version` and also console output when starting the game with `LD_DEBUG=libs`.

**Reply by Smorty100** [reply]
This is from _ffmpeg --version_ on my kde neon install (ubuntu based)
```
smorty@KDE:~$ ffmpeg --version
ffmpeg version 4.2.4-1ubuntu0.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.3.0-10ubuntu2)
  configuration: --prefix=/usr --extra-version=1ubuntu0.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
Unrecognized option '-version'.
Error splitting the argument list: Option not found
```

All of the log below is from the runtime log in the logs folder and comes from the appimage provided on github
And this is what I found in the runtime log when playing "this game" which should have a video in it:

```
2021-08-17 12:00:21 [verbose]: updating selection with beatmap:2288 ruleset:0
2021-08-17 12:00:21 [verbose]: beatmap changed from "Machico - Happy Magic (dkblaze) [Hard]" to "Suzuki Konomi - This game (TV Size) (Sekai-nyan) [Sekai's Hard Game]"
2021-08-17 12:00:21 [verbose]: working beatmap updated to Suzuki Konomi - This game (TV Size) (Sekai-nyan) [Sekai's Hard Game]
2021-08-17 12:00:28 [verbose]: Screen changed → PlayerLoader
2021-08-17 12:00:29 [verbose]: VideoDecoder faulted: System.DllNotFoundException: Unable to load shared library 'avformat' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libavformat: cannot open shared object file: No such file or directory
2021-08-17 12:00:29 [verbose]: at System.Runtime.InteropServices.NativeLibrary.LoadByName(String libraryName, QCallAssembly callingAssembly, Boolean hasDllImportSearchPathFlag, UInt32 dllImportSearchPathFlag, Boolean throwOnError)
2021-08-17 12:00:29 [verbose]: at System.Runtime.InteropServices.NativeLibrary.LoadLibraryByName(String libraryName, Assembly assembly, Nullable`1 searchPath, Boolean throwOnError)
2021-08-17 12:00:29 [verbose]: at System.Runtime.InteropServices.NativeLibrary.Load(String libraryName, Assembly assembly, Nullable`1 searchPath)
2021-08-17 12:00:29 [verbose]: at osu.Framework.Graphics.Video.VideoDecoder.<>c.<CreateFuncs>b__54_0(String name)
2021-08-17 12:00:29 [verbose]: at FFmpeg.AutoGen.ffmpeg.<>c.<.cctor>b__6_318()
2021-08-17 12:00:29 [verbose]: at FFmpeg.AutoGen.ffmpeg.avformat_alloc_context()
2021-08-17 12:00:29 [verbose]: at osu.Framework.Graphics.Video.VideoDecoder.prepareDecoding()
2021-08-17 12:00:29 [verbose]: at osu.Framework.Graphics.Video.VideoDecoder.StartDecoding()
2021-08-17 12:00:31 [verbose]: Screen changed → SoloPlayer
2021-08-17 12:02:01 [verbose]: Screen changed → SoloResultsScreen
```

I couldn't find a `LD_DEBUG=libs` in any log file.

**Reply by bdach** [reply]
`LD_DEBUG` is an environment variable. You're supposed to start the game executable with it set, as in:
```
$ LD_DEBUG=libs ./osu.AppImage
```
This _will_ be a dynamic linking failure as the log already shows, it's just a question of which paths the dynamic linker is trying to look in.

**Reply by Smorty100** [reply]
Alrighty. I tried piping the terminal output to the clipboard (the linux paste menu) but for some reason it just copied nothing. The output is so long, that my terminal actually cuts it off at the top. I started and exited the game and this is what I was beable to see in the terminal (the log.txt file)
[log.txt](https://github.com/ppy/osu/files/7000553/log.txt)

**Reply by bdach** [reply]
I see no trace of the game attempting to load FFmpeg in the above log. You need to actually attempt to start a map with video.

To capture output easier, you can redirect std{out,err} to a file directly:

```
$ LD_DEBUG=libs ./osu.AppImage > log.txt 2>&1
```

**Reply by Smorty100** [reply]
Okay, now the log has been saved correctly. During the test I played "this game" (again), finished it and once again no video showed up. There was also a message about the score not being able to submit, but I am assuming that this does not have anything to do with ffmpeg. The log is attached to this post.
[log.txt](https://github.com/ppy/osu/files/7009084/log.txt)

**Reply by bdach** [reply]
As evidenced by the log, the dynamic linker attempted to locate `libavformat` in the following locations:

```
     12592:	find library=libavformat.so [0]; searching
     12592:	 search cache=/etc/ld.so.cache
     12592:	 search path=/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib:/usr/lib		(system search path)
     12592:	  trying file=/lib/x86_64-linux-gnu/libavformat.so
     12592:	  trying file=/usr/lib/x86_64-linux-gnu/libavformat.so
     12592:	  trying file=/lib/libavformat.so
     12592:	  trying file=/usr/lib/libavformat.so
```

but was unable to find it there. It may be in one of the locations, but under a slightly different name, throwing off the dynamic linker. For example, on my machine, I have the following symlink set up:

```
/usr/lib/x86_64-linux-gnu$ ls -l | grep libavformat
lrwxrwxrwx  1 root root        24 lut 21 21:59 libavformat.so -> libavformat.so.58.45.100
```

You can create a similar symlink on your machine that links the FFmpeg libraries to one of the aforementioned locations for now. This in the long term will be resolved by bundling FFmpeg with the appimage (https://github.com/ppy/osu-framework/pull/3846).

---

Score submission failures are not related. Are you using the official appimages?
